clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -182,6 +215,222 @@ public CompletableFuture<Map<TopicPartition, OffsetAndTimestampInternal>> fetchO
+        // offsets according to the strategy (ex. earliest, latest), and update the positions.
+        resetPositionsIfNeeded().whenComplete((resetResult, error) -> {
+            if (error == null) {
+                result.complete(null);
+            } else {
+                result.completeExceptionally(error);
+            }
+        });
+    }
+
+    /**
+     * Fetch the committed offsets for partitions that require initialization. Use them to set
+     * the fetch positions in the subscription state.
+     *
+     * @throws TimeoutException If offsets could not be retrieved within the timeout
+     */
+    private CompletableFuture<Void> initWithCommittedOffsetsIfNeeded(long deadlineMs) {
+        final Set<TopicPartition> initializingPartitions = subscriptionState.initializingPartitions();
+
+        if (initializingPartitions.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        log.debug("Refreshing committed offsets for partitions {}", initializingPartitions);
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        // The shorter the timeout provided to poll(), the more likely the offsets fetch will time out. To handle
+        // this case, on the first attempt to fetch the committed offsets, a FetchCommittedOffsetsEvent is created
+        // (with potentially a longer timeout) and stored. The event is used for the first attempt, but in the
+        // case it times out, subsequent attempts will also use the event in order to wait for the results.
+        CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> fetchCommittedFuture;
+        if (!canReusePendingOffsetFetchEvent(initializingPartitions)) {
+            // Need to generate a new request to fetch committed offsets
+            final long fetchCommittedDeadlineMs = Math.max(deadlineMs, time.milliseconds() + defaultApiTimeoutMs);
+            fetchCommittedFuture = commitRequestManager.fetchOffsets(initializingPartitions, fetchCommittedDeadlineMs);
+            pendingOffsetFetchEvent = new PendingFetchCommittedRequest(initializingPartitions, fetchCommittedFuture);
+        } else {
+            fetchCommittedFuture = pendingOffsetFetchEvent.result;
+        }
+
+        // when the ongoing OffsetFetch completes, carry on with updating positions and
+        // completing the result future for the current attempt to initWithCommittedOffsetsIfNeeded
+        fetchCommittedFuture.whenComplete((offsets, error) -> {
+            pendingOffsetFetchEvent = null;
+
+            // Ensure we only set positions for the partitions that still require one (ex. some partitions may have
+            // been assigned a position manually)
+            if (error == null) {
+                Map<TopicPartition, OffsetAndMetadata> offsetsToApply = offsetsForInitializingPartitions(offsets);
+                refreshCommittedOffsets(offsetsToApply, metadata, subscriptionState);
```

Comment:
Is it legal to call `refreshCommittedOffsets`  multi times when there are many calls reusing the "first attempt"? i.e we register many actions to `whenComplete`

Reply:
good catch. I would say it's harmless because it has the check to be no-op if the partitions already got positions, but it's definitely conceptually wrong. We want to trigger the `OffsetFetch`, keep it as pending, and `refreshCommittedOffsets` only once when we get a response (what we need to do "multi times" is not the refreshCommittedOffsets, it's only to call "complete" on all the many attempts that may be waiting for that operation). I made the change.